util/group_manager.hpp

```diff
@@ -0,0 +1,174 @@
+  // if "update" has never successfully run. The returned command can be
+  // modified as desired before it is sent to the robot with the send function.
+  GroupCommand& pendingCommand() { return command_; }
+  const GroupCommand& pendingCommand() const { return command_; }
+
+  // Returns the last feedback obtained by update, or an empty feedback object
+  // if "update" has never successfully run.
+  const GroupFeedback& lastFeedback() const { return feedback_; }
+  const double dT() const { return dt_; }
+  const double lastTime() const { return last_time_; }
+
+  // Updates feedback.
+  // To retrieve the feedback, call `getLastFeedback()` after this call.
+  // You can modify the command object after calling this.
+  //
+  // Returns 'false' on a connection problem; true on success.
+  bool update();
+
+  // Sends the user set command to the group manager.
+  bool send();
+
+protected:
+  std::function<double()> get_current_time_s_;
+  double last_time_;
+  double dt_{ std::numeric_limits<double>::quiet_NaN() };
+  std::shared_ptr<Group> group_;
+
+  hebi::GroupFeedback feedback_;
+  hebi::GroupCommand command_;
+
+  // Private constructor
+  GroupManager(
+      std::function<double()> get_current_time_s,
+      std::shared_ptr<Group> group):
+    get_current_time_s_(get_current_time_s),
+    last_time_(get_current_time_s()),
+    group_(group),
+    feedback_(group->size()),
+    command_(group->size()) {}
+};
+
+std::unique_ptr<GroupManager> GroupManager::create(const GroupManager::Params& params) {
+
+  // Get the group (scope the lookup object so it is destroyed
+  // immediately after the lookup operation)
+  std::shared_ptr<Group> group;
+  {
+    Lookup lookup;
+    group = lookup.getGroupFromNames(params.families_, params.names_);
+  }
+  if (!group) {
+    std::cout << "Could not create arm! Check that family and names match actuators on the network.\n";
+    return nullptr;
+  }
+
+  // Set parameters
+  if (!group->setCommandLifetimeMs(params.command_lifetime_)) {
+    std::cout << "Could not set command lifetime on group; check that it is valid.\n";
+    return nullptr;
+  }
+  if (!group->setFeedbackFrequencyHz(params.control_frequency_)) {
+    std::cout << "Could not set feedback frequency on group; check that it is valid.\n";
+    return nullptr;
+  }
+
+  // Try to get feedback -- if we don't get a packet in the first N times,
+  // something is wrong
+  int num_attempts = 0;
+
+  // We need feedback, so we can plan trajectories if that gets called before the first "update"
+  GroupFeedback feedback(group->size());
+  while (!group->getNextFeedback(feedback)) {
+    if (num_attempts++ > 10) {
+      std::cout << "Could not communicate with robot; check network connection.\n";
+      return nullptr;
+    }
+  } 
+
+  // Note: once ROS moves up to C++14, we can change this to "make_unique".
+  return std::unique_ptr<GroupManager>(new GroupManager(params.get_current_time_s_, group));
```

Comment:
- create group manager earlier
- throw away `feedback` here
- use `_feedback` from `GroupManager` in the `getNextFeedback` call.
- return the created GroupManager on success.

Reply:
or -- you might be able to `std::move()` in the `feedback` and change the `GroupManager`'s constructor...